src/libs/safe-path/src/safe_dir_builder.rs

```diff
@@ -0,0 +1,154 @@
+    pub fn mode(&mut self, mode: u32) -> &mut Self {
+        self.mode = mode;
+        self
+    }
+
+    /// Sets the root directory for security check.
+    ///
+    /// # Safety
+    /// The `root` is the trusted part, and should not be modifiable by untrusted agents.
+    pub fn root<P: AsRef<Path>>(&mut self, root: P) -> &mut Self {
+        self.root = root.as_ref().to_path_buf();
+        self
+    }
+
+    /// Creates the specified directory with the options configured in this builder.
+    ///
+    /// The `path` must be a subdirectory of `SafePathBuf::root()`, otherwise error will be returned.
+    /// It is considered an error if the directory already exists unless recursive mode is enabled.
+    pub fn create<P: AsRef<Path>>(&self, path: P) -> Result<SafePathBuf> {
+        let mut root = self.root.canonicalize()?;
+        if !root.is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+
+        let path = safe_join("/", path)?;
+        let mut suffix = path.strip_prefix(&root).map_err(|_| {
+            Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", path.display()),
+            )
+        })?;
+        if suffix.file_name().is_none() {
+            return SafePathBuf::from_path(root);
+        }
+        if !self.recursive {
+            if let Some(parent) = path.parent() {
+                root = root.join(parent);
+            }
+            // Safe to unwrap() because we have verified `suffix` is not empty.
+            suffix = Path::new(suffix.file_name().unwrap());
+        }
+
+        for comp in suffix {
+            let file = SafePathBuf::from_path(&root)?;
+            if !file.target().is_dir() {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!("Invalid path: {}", root.display()),
+                ));
+            }
+            root = root.join(comp);
+            DirBuilder::new()
+                .mode(self.mode)
+                .recursive(true)
+                .create(&root)?;
+        }
+
+        let result = SafePathBuf::from_path(&root)?;
+        if !result.target().is_dir() {
+            return Err(Error::new(
+                ErrorKind::Other,
+                format!("Invalid path: {}", root.display()),
+            ));
+        }
+
+        Ok(result)
+    }
+}
+
+impl Default for SafeDirBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::fs;
+
+    #[test]
+    fn test_safe_path_buf() {
+        let rootfs_dir = tempfile::tempdir().expect("failed to create tmpdir");
+        let rootfs_path = rootfs_dir.path();
+        let mut builder = SafeDirBuilder::new();
+
+        builder.root(rootfs_path.join("__does_not_exist__"));
```

Comment:
Suggest using assert insteadof unwrap.

```rust
assert!(ok_value.is_ok());
assert!(eer_value.is_err());
```

Reply:
Any preference over this? I always use `unwrap()` in unit tests to unwrap the expected value.